### ISSUE TITLE
Rebuffi et al. - 2017 - iCaRL Incremental Classifier and Representation Learning

### ISSUE THREAD

**Issue by wagaken** [wagaken/PAPPER_SUMMARY#25]
CVPR 2017

**Comment by wagaken** [comment]
### Introduction
iCaRL, class incremental learning

**Problem setting**
Three properties that class-incremental learning should qualify 
1. it should be trainable from a stream of data in which examples of different classes occur at different times

2. it should be trainable from a stream of data in which examples of different classes occur at different times

3. its computational requirements and memory footprint should remain bounded, or at least grow very slowly, with respect to the number of classes seen so far.

The third criterion prevents trivial algorithms, such as storing all training examples and retraining an ordinary multi-class classifier whenever new data become available.

**Approach**
- classification of a nearest-mean-of-exemplars rule
- prioritized examplar selection based on herding
- representation learning using knowledge distillation and prototype rehearsal

**Characteristics**
only the training data for a small number of classes has to be present at the same time and new classes can be added progressively.

learns strong classifiers and a data representation simultaneously

### Method
<img width="362" alt="image" src="https://user-images.githubusercontent.com/49052500/155242854-0ae2c859-b7c7-49a4-8b8f-09511c8d3582.png">

- The mean-of-examlar classifier
<img width="359" alt="image" src="https://user-images.githubusercontent.com/49052500/155242280-6b2def6f-33a5-4428-837f-6844a5ad131e.png">

iCaRL ensures that the total number of exemplar images never exceeds a fixed parameter K.

<img width="374" alt="image" src="https://user-images.githubusercontent.com/49052500/155242758-910a8757-2ec5-47b5-aec7-38d690bb8f09.png">

<img width="358" alt="image" src="https://user-images.githubusercontent.com/49052500/155242810-f9e51388-0859-48d0-880f-32a4fa8b838c.png">


- Architecture
-iCaRL makes use of a CNN followed by a single classification layer with as many sigmoid output nodes as classes observed so far. All feature vectors are L2 normalized.

- Representation learning
<img width="362" alt="image" src="https://user-images.githubusercontent.com/49052500/155242905-390deb28-19ca-46ce-8a7f-d2d3a790f7e3.png">

First, iCaRL constructs an augmented training set consisting of the currently available training examples together with the stored exemplars. 

Next, the current network is evaluated for each example and the resulting network outputs for all previous classes are stored (not for the new classes, since the network has not been trained for these, yet). 

Finally, the network parameters are updated by minimizing a loss function that for
each new image encourages the network to output the correct class indicator for new classes (classification loss), and for old classes, to reproduce the scores stored in the previous step (distillation loss)

The representation learning resembles ordinary fine-tuning.
There are two modifications to plain fine-tuning.

First, the training set is augmented. It ensures that at least some information
about the data distribution of all previous classes enters the training process.

Second, the loss function is augmented. Besides the standard classification loss, which encourages improvements of the feature representation that allows classifying the newly observed classes well, it also contains the distillation loss, which ensures that the discriminative information learned previously is not lost during the new learning step

- Examplar management 
<img width="362" alt="image" src="https://user-images.githubusercontent.com/49052500/155243782-7fdfead0-4ab4-4900-a466-010e8be262c7.png">

<img width="355" alt="image" src="https://user-images.githubusercontent.com/49052500/155243526-7f8c268f-ac67-451a-a16d-04675984744a.png">

Whenever iCaRL encounters new classes it adjusts its exemplar set.
iCaRL will use m = K/t exemplars (up to rounding) for each class.
 The exemplar ”set” is really a prioritized list. The order of its elements matters, with exemplars earlier in the list being more important.

### Experimental result
<img width="314" alt="image" src="https://user-images.githubusercontent.com/49052500/155243894-3a84f104-48ff-40f3-a402-3310b4f86811.png">

<img width="311" alt="image" src="https://user-images.githubusercontent.com/49052500/155243938-53fb21c2-881f-4456-b240-9432004b3a89.png">
